test/controllers/ContactDetailsControllerSpec.scala

```diff
@@ -0,0 +1,122 @@
+/*
+ * Copyright 2019 HM Revenue & Customs
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package controllers
+
+import controllers.actions.{DataRetrievalAction, FakeActions}
+import forms.mappings.ContactDetailsMapping._
+import generators.Generators
+import models._
+import models.requests.SignedInUser
+import org.mockito.Mockito.{times, verify}
+import org.scalacheck.Arbitrary._
+import org.scalatest.concurrent.ScalaFutures
+import org.scalatest.mockito.MockitoSugar
+import org.scalatest.prop.PropertyChecks
+import pages.ContactDetailsPage
+import play.api.data.Form
+import play.api.libs.json.Json
+import play.api.test.Helpers._
+import uk.gov.hmrc.http.cache.client.CacheMap
+import views.html.contact_details
+
+class ContactDetailsControllerSpec extends ControllerSpecBase
+  with ScalaFutures
+  with MockitoSugar
+  with PropertyChecks
+  with Generators
+  with FakeActions {
+
+  val form = Form(contactDetailsMapping)
+
+  def view(form: Form[_] = form): String = contact_details(form)(fakeRequest, messages, appConfig).toString
+
+  def controller(signedInUser: SignedInUser, eori: String, dataRetrieval: DataRetrievalAction = getEmptyCacheMap) =
+    new ContactDetailsController(
+      messagesApi,
+      new FakeAuthAction(signedInUser),
+      new FakeEORIAction(eori),
+      dataRetrieval,
+      dataCacheConnector,
+      appConfig
+    )
+
+  "Contact details page" should {
+
+    "load the correct page when user is logged in" in {
+
+      forAll { (user: SignedInUser, eori: String) =>
+
+        val result = controller(user, eori).onPageLoad(fakeRequest)
+
+        status(result) mustBe OK
+        contentAsString(result) mustBe view(form)
+      }
+    }
+
+    "contact details should be displayed if it exist on the cache" in {
+
+      forAll { (user: SignedInUser, eori: String, contactDetails: ContactDetails) =>
+
+        val cacheMap: CacheMap = CacheMap("", Map(ContactDetailsPage.toString -> Json.toJson(contactDetails)))
+        val result = controller(user, eori, getCacheMap(cacheMap)).onPageLoad(fakeRequest)
+
+        contentAsString(result) mustBe view(form.fill(contactDetails))
+      }
+    }
+
+    "return an ok when valid data is submitted" in {
```

Comment:
The name of this test is wrong as it returns a `SEE_OTHER` and redirects the user
```suggestion
    "return an see other when valid data is submitted" in {
```